### DISCUSSION TITLE
Function score query is slow

### DISCUSSION THREAD

**Discussion by ArtemKar123** [alexklibisz/elastiknn#275] [Q&A] [answered]
I tried using function score query similar to what you can find in EKNN documentation, but it works 5 times slower than simple lsh query.
Any solution?

**Comment by alexklibisz** [comment]
Can you provide more detail about your setup, at least the following:
- Elastiknn version
- The LSH query
- The Function score query
- Index configuration (number of docs, number of shards)

**Reply by ArtemKar123** [reply]
I am using 7.13. There are 56 shards on 7 data-nodes and around 100 million docs.
The query is:
```
{
    "size": 5,
    "query":
    {
        "function_score": {
        "query": {
            "terms": { "color": colors }                  # colors= [1, 2, 3]
          },
          "functions": [                           # 2
            {
              "elastiknn_nearest_neighbors": {        
                  "field": "t_vector",                  
                  "vec": {                            
                      "values": list(vector),               
                  },
                  "candidates": 100, 
                  "model": "lsh", 
                  "probes": 3, 
                  "similarity": "angular"     
               } 
            }]
        }
      }
 ```

**Reply by alexklibisz** [reply]
Hmm, I don't see anything obviously wrong there. Can you try running with the [profile API](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-profile.html) to see where it's spending time? If that doesn't give an obvious answer, I would suggest attaching VisualVM to one of the elasticsearch nodes, setting it to sample CPU, and running a couple thousand queries to get an idea of where CPU is spent.

**Reply by ArtemKar123** [reply]
So, I tried to run an aggregation query like:
```
"aggregations": {
    "cid": {
      "terms": {
        "field": "color",
        "include": [1, 2]
      }
    }
```
   And it takes 350ms(which is more than a half of function score query above), while at the same time the terms query:
```
   "query": {
        "terms": {
          "color": [4, 5]
        }
  }
```
  takes only 60ms.
  
  I am pretty new to elastic, but I think the problem may be the aggregation for some reason.
  Is there any other method to run an elastiknn query on a subset of terms output?

**Reply by alexklibisz** [reply]
You can use a function score query or a query rescorer. Both of the patterns are here: https://elastiknn.com/api/#running-nearest-neighbors-query-on-a-filtered-subset-of-documents

I'm not really sure about aggregations. I haven't use them in a very long time.

**Reply by ArtemKar123** [reply]
There is a problem with rescoring too, because I have a large dataset I can not set window_size less than n_data/n_shards and for some reason it also works slower. 
I tried setting window_size = n_candidates*10, but it didn't return right results, because it would get the first 1000 of data in each shard and completely ignore other data.

**Reply by ArtemKar123** [reply]
It seems, I found what the problem is. It is in tests, I was applying. So, I was measuring time as mean of 200 queries, but when I made a simple search with ann, at some point elastic just cached the result and worked faster.
And with function scoring it is not able to cache any results, so it looks, like it is working slower, although it's not.

**Comment by alexklibisz** [comment, marked as answer]
> It seems, I found what the problem is. It is in tests, I was applying. So, I was measuring time as mean of 200 queries, but when I made a simple search with ann, at some point elastic just cached the result and worked faster.
And with function scoring it is not able to cache any results, so it looks, like it is working slower, although it's not.

That's interesting. Benchmarking Elasticsearch can be quite tricky, especially with such a large index.
